rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,162 @@
+
+## What
+
+Teleport should support trusting a third party OP and the JWTs that it issues when authenticating a client for the cluster join process. This is similar to the support for AWS IAM joining, and will allow joining a Teleport cluster without the need to distribute a static token on several platforms.
+
+Users will need to be able to configure trust in an OP, and rules that determine what identities are allowed to join the cluster.
+
+## Why
+
+This feature reduces the friction involved in dynamically adding many new nodes to Teleport on a variety of platforms. This is also more secure, as the user does not need to distribute a token which is liable to exfilitration.
+
+Whilst multiple providers offer OIDC identities to workloads running on their platform, we will start by targetting GCP GCE since this represents a key platform for Teleport, and is also well documented and easy to test on. However, the work towards this feature will also enable us to simply add other providers that support OIDC workload identity (see the references for more), this particularly ties into Machine ID goals as we aim to support several CI/CD providers that offer workload identity.
+
+## Details
+
+The work on OIDC joining is broken down into two parts:
+
+- Support in the Auth server for trusting an OP issued JWT. This work is general, and will be applicable to all providers.
+- Support in nodes for fetching their workload identity token from their environment. This work will be specific to each platform we intend to support.
+
+OIDC supports multiple types of token (`id_token`: a JWT encoding information about the identity, which can be verified using the issuer's public key and `access_token`: an opaque token that can be used with a `userinfo` endpoint on the issuer to obtain information about the identity). However, in the case of workload identities, `id_token` is the most prevelant. For this reason, our initial implementation will solely support `id_token`.
+
+### Auth server support
+
+We will re-use the existing endpoints around joining as much as possible. This means that the main entry-point for joining will be the existing `RegisterUsingToken` method.
+
+We will introduce a new token type, `oidc-jwt`, and add an additional field to the Token resource to allow the issuer URL to be specified (`issuer_url`).
+
+Registration flow:
+
+1. Client is configured by the user to use `oidc-jwt` joining with a specific provider. The client then uses the provider-specific logic to obtain a token.
+2. The client will call the `RegisterUsingToken` endpoint, specifying the name of the token, and providing the token that it has collected.
+3. The server will attempt to fetch the Token resource for the specified token.
+4. The server will check JWT header to ensure the `alg` is one we have whitelisted (RS256, TODO: Full list)
```

Comment:
I believe we will want to have an allow list of Algorithms here. There's been a number of historic vulnerabilities with JWT validation where engineers have allowed any algorithm to be used, and this has allowed less secure algorithms to be used, or in some cases even completely unsigned JWTs to pass validation.

Both Github and Google seem to use RS256, but for future-proofing, we will potentially want to support RS384 and RS512. We'll also want to support the other asymmetric options such as ECDSA. We'll want to prevent HMAC algorithms since these rely on a pre-shared key which is not used by OIDC.